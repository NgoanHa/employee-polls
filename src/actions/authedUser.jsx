export const SET_AUTHENTICATION_USER = "SET_AUTHENTICATION_USER";
export const LOGOUT_AUTHENTICATION_USER = "LOGOUT_AUTHENTICATION_USER";

export const setAuthenticationUser = (authedUser) => {
    return {
        type: SET_AUTHENTICATION_USER,
        authedUser
    }
}

export const logoutAuthenticationUser = () => {
    return {
        type: LOGOUT_AUTHENTICATION_USER
    }
}

export const hanldeLoginProcess = (username, password) => {
    return (dispatch, getState) => {
        const { users } = getState();
        const user = Object.values(users).find((user) => user.id === username && user.password === password);
        if(!!user){
            return dispatch(setAuthenticationUser(user));
        }
    }
}

export const handleLogoutProcess = () => {
    return (dispatch) => {
        return dispatch(logoutAuthenticationUser());
    };
}