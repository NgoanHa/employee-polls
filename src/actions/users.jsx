export const GET_USERS = "GET_USERS";
export const ADD_ANSWER_USER = "ADD_ANSWER_USER";
export const ADD_QUESTION_USER = "ADD_QUESTION_USER";

export const getUsers = (users) => {
    return {
        type: GET_USERS,
        users
    }
}

export const addAnswerOfUser = (authedUser, id, answer) => {
    return {
        type: ADD_ANSWER_USER,
        authedUser,
        qId: id,
        answer,
    }
}

export const addQuestionOfUser = (authedUser, id) => {
    return {
        type: ADD_QUESTION_USER,
        authedUser,
        qid: id,
    }   
}