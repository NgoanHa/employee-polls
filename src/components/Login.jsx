import { Navigate } from 'react-router-dom';
import { connect } from 'react-redux'
import { useState } from 'react';
import { hanldeLoginProcess } from '../actions/authedUser';
import '../style/common.css'

const Login = ({dispatch, logInState}) => {

    const [username, setUsername] = useState("sarahedo");
    const [password, setPassword] = useState("sarahedo123");

    if(logInState){
        return <Navigate to="/"/>;
    }

    const handleSetUsername = (e) => {
        const usernameValue = e.target.value;
        setUsername(usernameValue);
    };

    const handleSetPassword = (e) => {
        const passwordValue = e.target.value;
        setPassword(passwordValue);
    };

    const handleFormSubmit = (e) => {
        e.preventDefault();
        dispatch(hanldeLoginProcess(username, password));
        setUsername("");
        setPassword("");
    };

    return (
        <div className="login-container">
            <h1>Log In</h1>
            <form className="login-container-form" onSubmit={handleFormSubmit}>
                <div className="login-container-username">
                    <label htmlFor="username">Username</label>
                    <input type="text" name="username" id="username" data-testid="username" value={username} onChange={handleSetUsername}/>
                </div>
                <div className="login-container-password">
                    <label htmlFor="password">Password</label>
                    <input type="password" name="password" id="password" data-testid="password" value={password} onChange={handleSetPassword}/>
                </div>
                <div>
                    <button data-testid="submit" type="submit" className="login-container-btn">Submit</button>
                </div>
            </form>
        </div>
    );
}

const mapStateToProps = ({authedUser}) => ({
    logInState: !!authedUser,
});
export default connect(mapStateToProps)(Login);