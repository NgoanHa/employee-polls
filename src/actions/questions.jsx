import { saveQuestion, saveQuestionAnswer } from "../util/HandleServerData";
import { addQuestionOfUser, addAnswerOfUser } from "./users";

//Define constant type
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";
export const GET_QUESTIONS = "GET_QUESTIONS";

export const getQuestions = (questions) => {
    return {
        type: GET_QUESTIONS,
        questions
    }
}

export const addQuestion = (question) => {
    return {
        type: ADD_QUESTION,
        question
    }
}

export const addAnswerQuestion = (author, qid, answer) => {
    return {
        type: ADD_ANSWER_QUESTION,
        author,
        qid,
        answer
    }
}

export const handleAddQuestion = (optionOneText, optionTwoText) => {
    return async (dispatch, getState) => {
        const { authedUser } = getState();
        return await saveQuestion(optionOneText, optionTwoText, authedUser.id).then((question) => {
            dispatch(addQuestion(question));
            dispatch(addQuestionOfUser(authedUser.id));
        })
    }
}

export const handleAddAnswer = (qid, answer) => {
    return async (dispatch, getState) => {
        const { authedUser } = getState();
        return await saveQuestionAnswer(authedUser.id, qid, answer).then(() => {
            dispatch(addAnswerQuestion(authedUser.id, qid, answer));
            dispatch(addAnswerOfUser(authedUser.id, qid, answer));
        });

    }
}