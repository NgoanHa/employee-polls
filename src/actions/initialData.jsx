import { getInitData } from "../util/HandleServerData"
import { getQuestions } from "./questions"
import { getUsers } from "./users"

export const handleInitData = () => {
    return (dispatch) => {
        return getInitData().then(({users, questions}) => {
            dispatch(getUsers(users));
            dispatch(getQuestions(questions));
        });     
    };
}