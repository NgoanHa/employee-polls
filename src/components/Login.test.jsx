import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../store";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import Login from "./Login";
import { handleInitData } from '../actions/initialData';

describe("Login", () => {
    it("login component is render", () => {
        const actual = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        );
        expect(actual).toBeDefined();
        expect(actual).toMatchSnapshot();
    });

    it('login with invalid password', async () => {
        await store.dispatch(handleInitData());

        const actual = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Login />
                </BrowserRouter>
            </Provider>
        );

        const username = actual.getByTestId("username");
        const password = actual.getByTestId("password");
        const submitBtn = actual.getByTestId("submit");
        expect(username).toBeInTheDocument();
        expect(password).toBeInTheDocument();
        expect(submitBtn).toBeInTheDocument();

        fireEvent.change(username, { target: { value: 'sarahedo' } });
        fireEvent.change(password, { target: { value: 'invalidpassword' } });
        expect(username.value).toBe("sarahedo");
        expect(password.value).toBe("invalidpassword");
        fireEvent.click(submitBtn);
        expect(username.value).toBe("");
        expect(password.value).toBe("");
    });
});