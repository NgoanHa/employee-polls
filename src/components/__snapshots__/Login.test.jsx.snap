// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login login component is render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-container"
      >
        <h1>
          Log In
        </h1>
        <form
          class="login-container-form"
        >
          <div
            class="login-container-username"
          >
            <label
              for="username"
            >
              Username
            </label>
            <input
              data-testid="username"
              id="username"
              name="username"
              type="text"
              value="sarahedo"
            />
          </div>
          <div
            class="login-container-password"
          >
            <label
              for="password"
            >
              Password
            </label>
            <input
              data-testid="password"
              id="password"
              name="password"
              type="password"
              value="sarahedo123"
            />
          </div>
          <div>
            <button
              class="login-container-btn"
              data-testid="submit"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-container"
    >
      <h1>
        Log In
      </h1>
      <form
        class="login-container-form"
      >
        <div
          class="login-container-username"
        >
          <label
            for="username"
          >
            Username
          </label>
          <input
            data-testid="username"
            id="username"
            name="username"
            type="text"
            value="sarahedo"
          />
        </div>
        <div
          class="login-container-password"
        >
          <label
            for="password"
          >
            Password
          </label>
          <input
            data-testid="password"
            id="password"
            name="password"
            type="password"
            value="sarahedo123"
          />
        </div>
        <div>
          <button
            class="login-container-btn"
            data-testid="submit"
            type="submit"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
